File: rl_2048/ntuple/train.py
"""Training loop for N-tuple networks."""

import logging
from collections import Counter
from dataclasses import asdict
from datetime import datetime
from pathlib import Path

from torch.utils.tensorboard import SummaryWriter

from rl_2048.game import Action, Board, Game2048, apply_action
from rl_2048.ntuple.config import NTupleConfig
from rl_2048.ntuple.network import NTupleNetwork
from rl_2048.profiler import Profiler

logger = logging.getLogger(__name__)


def train(
    config: NTupleConfig,
    run_dir: str = "runs",
    run_name: str | None = None,
    resume: str | None = None,
    offset: int = 0,
):
    """
    Main n-tuple network training loop.

    Parameters
    ----------
    resume : str | None
        Path to a checkpoint .npz file to resume training from
    offset : int
        Episode offset to start from when resuming
    """
    logger.info("Training started")

    group_name = run_name or "ntuple"
    timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
    run_path = Path(run_dir) / group_name / timestamp

    writer = SummaryWriter(log_dir=str(run_path))
    logger.info("TensorBoard log dir: %s", writer.log_dir)
    hparams_subdir = _log_hyperparams(writer, config)
    checkpoint_path = run_path / hparams_subdir / "checkpoints"
    checkpoint_path.mkdir(parents=True, exist_ok=True)

    game = Game2048()

    if resume:
        resume_path = Path(resume)
        network = NTupleNetwork.load(resume_path)
        logger.info("Loaded checkpoint from %s", resume_path)
    else:
        network = NTupleNetwork(config.patterns, v_init=config.v_init)

    _log_network(writer, network)
    profiler = Profiler()

    last_eval: dict = {}

    try:
        for episode in range(1 + offset, config.max_episodes + 1):
            _run_episode(game, network, config, profiler)

            max_tile = max(game.board)
            writer.add_scalar("train/score", game.score, episode)
            writer.add_scalar("train/max_tile", max_tile, episode)

            if episode % 100 == 0:
                logger.info(
                    "Ep %d | Score: %d | Max tile: %d",
                    episode,
                    game.score,
                    max_tile,
                )
                profiler.log_and_reset(episode, writer)

            if episode % config.eval_interval == 0:
                profiler.begin()
                last_eval = evaluate(game, network, config.eval_episodes)
                profiler.record("eval")
                logger.info(
                    "EVAL Ep %d | Mean score: %.0f | Max: %d | Tiles: %s",
                    episode,
                    last_eval["mean_score"],
                    last_eval["max_score"],
                    last_eval["tile_distribution"],
                )
                writer.add_scalar("eval/mean_score", last_eval["mean_score"], episode)
                writer.add_scalar("eval/max_score", last_eval["max_score"], episode)
                for tile_val, count in last_eval["tile_distribution"].items():
                    pct = count / config.eval_episodes
                    writer.add_scalar(f"eval/tile_pct{tile_val}", pct, episode)

                profiler.begin()
                ep_str = str(episode).zfill(len(str(config.max_episodes)))
                network.save(checkpoint_path / f"checkpoint_ep{ep_str}.npz")
                profiler.record("checkpoint_save")

        network.save(checkpoint_path / "final.npz")
        logger.info(
            "Training complete. Model saved to %s",
            checkpoint_path / "final.npz",
        )
    finally:
        writer.add_scalar("final/mean_score", last_eval.get("mean_score", 0.0))
        writer.add_scalar("final/max_score", last_eval.get("max_score", 0))
        writer.close()


def _run_episode(
    game: Game2048,
    network: NTupleNetwork,
    config: NTupleConfig,
    profiler: Profiler,
):
    """Play one episode with online TD(0) updates."""
    game.reset()

    while True:
        profiler.begin()
        action, afterstate = _select_action(network, game.board)
        profiler.record("select_action")
        if action is None or afterstate is None:
            break

        current_value = network.evaluate(afterstate)

        profiler.begin()
        game.step(action)
        profiler.record("env_step")

        # TD target from post-spawn board
        profiler.begin()
        target = _best_afterstate_value(network, game.board)
        td_error = target - current_value
        network.update(afterstate, config.lr * td_error)
        profiler.record("td_update")


def evaluate(game: Game2048, network: NTupleNetwork, num_episodes: int) -> dict:
    """Run evaluation episodes with greedy policy (no updates)."""
    scores: list[int] = []
    max_tiles: list[int] = []
    for _ in range(num_episodes):
        game.reset()
        while True:
            action, _ = _select_action(network, game.board)
            if action is None:
                break
            game.step(action)
        scores.append(game.score)
        max_tiles.append(max(game.board))

    tile_counts = Counter(max_tiles)
    return {
        "mean_score": sum(scores) / len(scores),
        "max_score": max(scores),
        "tile_distribution": dict(sorted(tile_counts.items())),
    }


def _best_afterstate_value(network: NTupleNetwork, board: Board) -> float:
    """
    Compute max_a [reward(a) + V(afterstate(a))] from a board state. Returns 0 if no
    valid actions (terminal).
    """
    best = 0.0
    found_valid = False
    for action in Action:
        afterstate, reward = apply_action(board, action)
        if afterstate == board:
            continue
        value = reward + network.evaluate(afterstate)
        if not found_valid or value > best:
            best = value
            found_valid = True
    return best


def _select_action(
    network: NTupleNetwork,
    board: Board,
) -> tuple[Action | None, Board | None]:
    """Select best action, returning (action, afterstate). (None, None) if terminal."""
    best_action: Action | None = None
    best_afterstate: Board | None = None
    best_value = float("-inf")

    for action in Action:
        afterstate, reward = apply_action(board, action)
        if afterstate == board:
            continue
        value = reward + network.evaluate(afterstate)
        if value > best_value:
            best_action = action
            best_afterstate = afterstate
            best_value = value

    return best_action, best_afterstate


def _log_hyperparams(writer: SummaryWriter, config: NTupleConfig) -> str:
    """Log hyperparameters for TensorBoard. Returns hparams subdirectory name."""
    hparam_dict = {
        k: v
        for k, v in asdict(config).items()
        if isinstance(v, (int, float, str, bool))
    }
    hparam_dict["num_patterns"] = len(config.patterns)
    hparam_dict["tuple_size"] = len(config.patterns[0]) if config.patterns else 0
    metric_dict = {
        "final/mean_score": 0.0,
        "final/max_score": 0,
    }
    run_name = datetime.now().strftime("%Y%m%d_%H%M%S")
    writer.add_hparams(hparam_dict, metric_dict, run_name=run_name)
    return run_name


def _log_network(writer: SummaryWriter, network: NTupleNetwork):
    """Log network info to TensorBoard."""
    total_entries = sum(len(lut) for lut in network.luts)
    memory_mb = total_entries * 4 / (1024 * 1024)
    info = (
        f"N-Tuple Network: {network.num_patterns} patterns, "
        f"{total_entries:,} total LUT entries, "
        f"{memory_mb:.0f} MB"
    )
    logger.info(info)
    writer.add_text("network/architecture", info)
    writer.add_text(
        "network/patterns",
        "\n".join(f"Pattern {i}: {p}" for i, p in enumerate(network.patterns)),
    )

File: rust_2048/src/train.rs
/// Training loop for N-tuple network with TD(0) afterstate learning.

use std::collections::HashMap;
use std::path::Path;
use std::time::Instant;

use crate::board::{Action, Board, Game, apply_action, max_tile};
use crate::ntuple::NTupleNetwork;

pub struct TrainConfig {
    pub lr: f32,
    pub max_episodes: u32,
    pub eval_interval: u32,
    pub eval_episodes: u32,
    pub v_init: f32,
    pub checkpoint_dir: String,
}

impl Default for TrainConfig {
    fn default() -> Self {
        Self {
            lr: 0.0025,
            max_episodes: 100_000,
            eval_interval: 1_000,
            eval_episodes: 25,
            v_init: 0.0,
            checkpoint_dir: "checkpoints".into(),
        }
    }
}

/// Select the best action: argmax_a [reward(a) + V(afterstate(a))].
/// Returns (action, afterstate) or None if terminal.
#[inline]
pub fn select_action(network: &NTupleNetwork, board: Board) -> Option<(Action, Board)> {
    let mut best_action: Option<Action> = None;
    let mut best_afterstate: Board = 0;
    let mut best_value = f64::NEG_INFINITY;

    for &action in &Action::ALL {
        let (afterstate, reward) = apply_action(board, action);
        if afterstate == board {
            continue;
        }
        let value = reward as f64 + network.evaluate(afterstate) as f64;
        if value > best_value {
            best_action = Some(action);
            best_afterstate = afterstate;
            best_value = value;
        }
    }

    best_action.map(|a| (a, best_afterstate))
}

/// Compute max_a [reward(a) + V(afterstate(a))] from a board state.
/// Returns 0 if terminal (no valid actions).
#[inline]
fn best_afterstate_value(network: &NTupleNetwork, board: Board) -> f32 {
    let mut best = 0.0f64;
    let mut found = false;

    for &action in &Action::ALL {
        let (afterstate, reward) = apply_action(board, action);
        if afterstate == board {
            continue;
        }
        let value = reward as f64 + network.evaluate(afterstate) as f64;
        if !found || value > best {
            best = value;
            found = true;
        }
    }

    if found { best as f32 } else { 0.0 }
}

/// Run one episode with online TD(0) updates.
fn run_episode(game: &mut Game, network: &mut NTupleNetwork, lr: f32) {
    game.reset();

    loop {
        let Some((action, afterstate)) = select_action(network, game.board) else {
            break;
        };

        let current_value = network.evaluate(afterstate);

        game.step(action);

        let target = best_afterstate_value(network, game.board);
        let td_error = target - current_value;
        network.update(afterstate, lr * td_error);
    }
}

/// Evaluate the network over multiple episodes with greedy policy (no updates).
fn evaluate(game: &mut Game, network: &NTupleNetwork, num_episodes: u32) -> EvalResult {
    let mut scores = Vec::with_capacity(num_episodes as usize);
    let mut max_tiles = Vec::with_capacity(num_episodes as usize);

    for _ in 0..num_episodes {
        game.reset();
        loop {
            let Some((action, _)) = select_action(network, game.board) else {
                break;
            };
            game.step(action);
        }
        scores.push(game.score);
        max_tiles.push(max_tile(game.board));
    }

    let mean_score = scores.iter().map(|&s| s as f64).sum::<f64>() / scores.len() as f64;
    let max_score = *scores.iter().max().unwrap_or(&0);

    let mut tile_distribution: HashMap<u32, u32> = HashMap::new();
    for &t in &max_tiles {
        *tile_distribution.entry(t).or_insert(0) += 1;
    }

    EvalResult {
        mean_score,
        max_score,
        tile_distribution,
    }
}

struct EvalResult {
    mean_score: f64,
    max_score: u32,
    tile_distribution: HashMap<u32, u32>,
}

impl std::fmt::Display for EvalResult {
    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
        let mut tiles: Vec<_> = self.tile_distribution.iter().collect();
        tiles.sort_by_key(|(k, _)| *k);
        let tile_str: Vec<String> = tiles.iter().map(|(k, v)| format!("{k}: {v}")).collect();
        write!(
            f,
            "mean={:.0}  max={}  tiles={{{}}}",
            self.mean_score,
            self.max_score,
            tile_str.join(", ")
        )
    }
}

/// Extract the episode number from a checkpoint filename like "checkpoint_ep0100000.bin".
fn parse_checkpoint_episode(path: &str) -> Option<u32> {
    let filename = Path::new(path).file_stem()?.to_str()?;
    let ep_str = filename.strip_prefix("checkpoint_ep")?;
    ep_str.parse().ok()
}

/// Main training entry point.
pub fn train(config: &TrainConfig, resume: Option<&str>) {
    let mut game = Game::new();

    let start_episode = resume
        .and_then(parse_checkpoint_episode)
        .unwrap_or(0);

    let mut network = if let Some(path) = resume {
        eprintln!("Resuming from {path} (episode {start_episode})");
        NTupleNetwork::load(Path::new(path)).expect("Failed to load checkpoint")
    } else {
        NTupleNetwork::new(crate::ntuple::DEFAULT_PATTERNS, config.v_init)
    };

    let checkpoint_dir = Path::new(&config.checkpoint_dir);
    std::fs::create_dir_all(checkpoint_dir).ok();

    let end_episode = start_episode + config.max_episodes;
    let width = (end_episode as f64).log10().ceil() as usize;
    let start_time = Instant::now();

    eprintln!(
        "Training episodes {}..{}  lr={}  eval_interval={}  eval_episodes={}  v_init={}",
        start_episode + 1, end_episode, config.lr, config.eval_interval, config.eval_episodes, config.v_init
    );

    for episode in (start_episode + 1)..=end_episode {
        run_episode(&mut game, &mut network, config.lr);

        if episode % config.eval_interval == 0 {
            let elapsed = start_time.elapsed().as_secs_f64();
            let trained = episode - start_episode;
            let eps = trained as f64 / elapsed;
            let result = evaluate(&mut game, &network, config.eval_episodes);
            eprintln!(
                "EVAL ep {:>width$}  {}  [{:.1}s, {:.0} ep/s]",
                episode, result, elapsed, eps,
            );

            let filename = format!("checkpoint_ep{:0>width$}.bin", episode);
            let path = checkpoint_dir.join(filename);
            if let Err(e) = network.save(&path) {
                eprintln!("  Warning: failed to save checkpoint: {e}");
            }
        }
    }

    let final_path = checkpoint_dir.join("final.bin");
    network.save(&final_path).expect("Failed to save final checkpoint");

    let total_time = start_time.elapsed().as_secs_f64();
    eprintln!(
        "Training complete. {} episodes in {:.1}s ({:.0} ep/s). Saved to {:?}",
        config.max_episodes,
        total_time,
        config.max_episodes as f64 / total_time,
        final_path,
    );
}
